### PR TITLE
[AutoDiff] Revamp `PullbackInfo`

### PR DESCRIPTION
**Pull request by bartchr808**
## `PullbackInfo` -> `LinearMapInfo`
With the introduction of Forward Mode Differentiation, the name `PullbackInfo` is not correct anymore. To recall, it stores information about the JVP/VJP generation. thus, a more general name of `LinearMapInfo` more appropriately represents what it stores. 

Specifically, it stores the differential and pullback functions, which are linear functions. We map those functions to the original apply calls, which is where `LinearMapInfo` and `linearMapStructs` comes from.

## Revamp `linearMapStruct` creation to occur before JVP/VJP emission
In my forward mode PR #26057 , I experienced an LLVM IR bug. What was going on is that it was caching the linear map structs before the linear functions were added. Those functions were being added while we were visiting the relevant instructions in JVP/VJP emission. 

The solution to avoid this caching problem was to bring the part where we add the linear map functions earlier to when we create the `LinearMapInfo`. Thus, when the JVP/VJP emission pass occurs, it will cache the full linear map struct.

### CODE REVIEW

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -1357,6 +1398,278 @@ class DifferentiableActivityInfo {
...
+  // should be differentiated.
+  if (ai->mayWriteToMemory())
+    for (auto &op : ai->getAllOperands())
+      if (activityInfo.isActive(op.get(), indices)) {
+        return true;
+      }
+  return false;
+}
+
+ void LinearMapInfo::prepareLinearMapStructDeclarations(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Indentation of this entire function seems largely messed up.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -1357,6 +1398,278 @@ class DifferentiableActivityInfo {
...
+                ai->getArgumentsWithoutIndirectResults().size(),
+                activeParamIndices);
+          }
+          SILAutoDiffIndices curIndices(activeResultIndices.front(),
+          AutoDiffIndexSubset::get(
+              builder.getASTContext(),
+              ai->getArgumentsWithoutIndirectResults().size(),
+              activeParamIndices));
+
+           // Check and diagnose non-differentiable original function type.
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: This is not diagnosing. Please update names to match its new semantics when reorganizing code.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -1357,6 +1398,278 @@ class DifferentiableActivityInfo {
...
+                        .getSILStorageType()
+                        .isDifferentiable(builder.getModule())) {
+                  return true;
+                }
+                return false;
+              };
+          if (diagnoseNondifferentiableOriginalFunctionType(originalFnSubstTy))
+            continue;
+
+           auto JVPType = originalFnSubstTy->getAutoDiffAssociatedFunctionType(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Camel case for all.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 4]
```diff
@@ -1357,6 +1398,278 @@ class DifferentiableActivityInfo {
...
+    branchingTraceDecls.insert({&origBB, traceEnum});
+    if (origBB.isEntry())
+      continue;
+    auto *traceEnumField =
+        addVarDecl(linearMapStruct, astCtx.getIdentifier("predecessor").str(),
+                   traceEnum->getDeclaredInterfaceType());
+    linearMapStructEnumFields.insert({linearMapStruct, traceEnumField});
+  }
+
+   // Add the differential function fields to the differential structs.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Comment should not be "differential"-specific.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 3 above)
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
           auto assocFnType = originalFnSubstTy->getAutoDiffAssociatedFunctionType(
```
`JVPXXX` -> `assocFnXXX` and `differentialXXX` -> `linearMapXXX` below as well please.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 5]
```diff
@@ -466,210 +480,269 @@ class PullbackInfo {
...
+    case swift::AutoDiffAssociatedFunctionKind::VJP:
+      linearMapName = "pullback_" + llvm::itostr(linearMapValueMap.size());
+      break;
+    }
+    auto *linearMapDecl = addVarDecl(linMapStruct, linearMapName, astFnTy);
+    linearMapValueMap.insert({inst, linearMapDecl});
+    return linearMapDecl;
   }
 
+  void prepareLinearMapStructDeclarations(
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Please add a doc comment clarifying that this method populates the fields of linear map structs. Or perhaps rename the method to something like `populateLinearMapStructDeclarationFields`.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 6]
```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
...
+            ai->getArgumentsWithoutIndirectResults()[i], indices)) {
+      return true;
+    }
+  // Instructions that may write to memory and that have an active operand
+  // should be differentiated.
+  if (ai->mayWriteToMemory())
+    for (auto &op : ai->getAllOperands())
+      if (activityInfo.isActive(op.get(), indices)) {
+        return true;
+      }
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Remove `{}`.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 7]
```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
...
+      return true;
+    }
+  // Function applications with an inout argument should be differentiated.
+  auto paramInfos = ai->getSubstCalleeConv().getParameters();
+  for (auto i : swift::indices(paramInfos))
+    if (paramInfos[i].isIndirectInOut() &&
+        activityInfo.isActive(
+            ai->getArgumentsWithoutIndirectResults()[i], indices)) {
+      return true;
+    }
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Remove {}

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 8]
```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
...
+    return activityInfo.isActive(val, indices);
+  })) {
+    return true;
+  }
+  // Function applications with an active indirect result should be
+  // differentiated.
+  for (auto indRes : ai->getIndirectSILResults())
+    if (activityInfo.isActive(indRes, indices)) {
+      return true;
+    }
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Remove {}.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 9]
```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
...
+  // should be differentiated.
+  if (ai->mayWriteToMemory())
+    for (auto &op : ai->getAllOperands())
+      if (activityInfo.isActive(op.get(), indices)) {
+        return true;
+      }
+  return false;
+}
+
+ void LinearMapInfo::populateLinearMapStructDeclarationFields(
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
void LinearMapInfo::populateLinearMapStructDeclarationFields(
```

Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I'm fine with checking this function to unblock things, but it should be broken down into multiple functions to reduce nesting.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 10]
```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
...
+            !ai->hasSemantics("array.uninitialized_intrinsic")) {
+          SmallVector<SILValue, 4> allResults;
+          allResults.push_back(ai);
+          allResults.append(ai->getIndirectSILResults().begin(),
+                            ai->getIndirectSILResults().end());
+
+          // Check if there are any active results or arguments. If not, skip
+          // this instruction.
+          auto hasActiveResults = llvm::any_of(
+             allResults, [&](SILValue res) {
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
              allResults, [&](SILValue res) {
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 11]
```diff
@@ -1357,6 +1402,285 @@ class DifferentiableActivityInfo {
...
+            return activityInfo.isActive(res, indices);
+          });
+          auto hasActiveArguments = llvm::any_of(
+             ai->getArgumentsWithoutIndirectResults(), [&](SILValue arg) {
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
              ai->getArgumentsWithoutIndirectResults(), [&](SILValue arg) {
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 6 above)
Comment (bartchr808) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: There are some other places I can add this simplification to below in my code, will do it for those one liner `return true` as well

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 9 above)
Comment (bartchr808) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: No worries I can make it part of this PR. Currently rebuilding my Xcode build, so will push a separate commit to make this change soon once it finishes.

Comment (bartchr808) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Added this change in my latest PR, separated out the code of when I find an apply instruction.

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 9 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks!

**Pull request merged**